### PR TITLE
RFC: wharf-api migrations

### PR DESCRIPTION
**Pull request by applejag**
[GORM's auto migrations](https://gorm.io/docs/migration.html#Auto-Migration) is great for simple migrations, but doesn't allow us to add custom migration logic. This RFC proposes a way to add on to our migration flow to both not run migrations when they're not needed as well as allow us to add any custom logic, such as concatenating two columns together, or as in [wharf-api#133](https://github.com/iver-wharf/wharf-api/issues/133)'s case that needs to parse test result artifacts again.

## Rendered

https://github.com/iver-wharf/rfcs/blob/rfc/wharf-api-migrations/_published/0029-wharf-api-migrations.md

### CODE REVIEW

**Comment by fredx30** [comment]
Consider checking if either of these public projects may be worth implementing rather than writing our own code for this.
https://github.com/go-gormigrate/gormigrate#gormigrate
https://github.com/golang-migrate/migrate

There are likely more external options to be considered if these are not relevant if found these during a 2 minute google search. I strongly advocate for using a standard thats out there, polished, and popular rather than inventing our own.

**Comment by applejag** [comment]
> @fredx30 
> Consider checking if either of these public projects may be worth implementing rather than writing our own code for this. https://github.com/go-gormigrate/gormigrate#gormigrate https://github.com/golang-migrate/migrate
> 
> There are likely more external options to be considered if these are not relevant if found these during a 2 minute google search. I strongly advocate for using a standard thats out there, polished, and popular rather than inventing our own.

Yea I feel kind of embarassed for not looking. Maybe I got lost in the moment or something. The gormigrate you found looks really promising, but I'll also spend some time digging.

File: _published/0029-wharf-api-migrations.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+1. `VERSION`: wharf-api version with `v` prefix.
+
+2. `YYYY-MM-DD`: date in format of year-month-day, with month and day being
+   left-padded with zeros.
+
+3. `hh:mm:ss`: time in format of hour:minute:seconds, with hour ranging from
+   00-23, and all being left-padded with zeros.
+
+All above being values relative to when the migrations were written by the
+developer. The date and time shall be is in UTC.
```
Comment (Pikabanga) [inline _published/0029-wharf-api-migrations.md]:
```suggestion
developer. The date and time shall be in UTC.
```

**Review by Pikabanga — APPROVED** [review]

**Comment by Pikabanga** [comment]
Does Gormigrate support custom SQL to be executed for complex scenarios like data transformation, table splitting etc?

I'm sure I can find out myself but I thought I'd be lazy since I guess you already know the answer to this :)

**Comment by applejag** [comment]
> @Pikabanga 
> Does Gormigrate support custom SQL to be executed for complex scenarios like data transformation, table splitting etc?
> 
> I'm sure I can find out myself but I thought I'd be lazy since I guess you already know the answer to this :)

Yes it does, via GORM. Docs: https://gorm.io/docs/sql_builder.html

In the migration definition, seen in https://github.com/iver-wharf/rfcs/blob/rfc/wharf-api-migrations/_published/0029-wharf-api-migrations.md#writing-migrations, we get an argument of `tx *gorm.DB`. The `.Raw()` from the docs is a method on `*gorm.DB`.

**Edit:** It's called `tx` in the migration as a conventional abbreviation of "transaction", meaning it's a database object that operates inside a transaction.

**Comment by Pikabanga** [comment]
> > @Pikabanga
> > Does Gormigrate support custom SQL to be executed for complex scenarios like data transformation, table splitting etc?
> > I'm sure I can find out myself but I thought I'd be lazy since I guess you already know the answer to this :)
> 
> Yes it does, via GORM. Docs: https://gorm.io/docs/sql_builder.html
> 
> In the migration definition, seen in https://github.com/iver-wharf/rfcs/blob/rfc/wharf-api-migrations/_published/0029-wharf-api-migrations.md#writing-migrations, we get an argument of `tx *gorm.DB`. The `.Raw()` from the docs is a method on `*gorm.DB`.
> 
> **Edit:** It's called `tx` in the migration as a conventional abbreviation of "transaction", meaning it's a database object that operates inside a transaction.

Perfect!

**Review by Pikabanga — APPROVED** [review]

**Review by Alexamakans — APPROVED** [review]

**Pull request merged**